harmony/http.py

```diff
@@ -249,28 +252,46 @@ def download(config, url: str, access_token: str, data, destination_file):
 
     response = None
     logger = build_logger(config)
-    logger.info('Downloading %s', url)
+    start_time = datetime.datetime.now()
+    logger.info(f'timing.download.start {url}')
 
     if data is not None:
         logger.info('Query parameters supplied, will use POST method.')
         data = urlencode(data).encode('utf-8')
 
     if access_token is not None and _valid(config.oauth_host, config.oauth_client_id, access_token):
         response = _download(config, url, access_token, data)
-        if response.ok:
-            destination_file.write(response.content)
-            logger.info(f'Completed {url}')
-            return response
-
-    if config.fallback_authn_enabled:
-        msg = ('No valid user access token in request or EDL OAuth authentication failed.'
-               'Fallback authentication enabled: retrying with Basic auth.')
-        logger.warning(msg)
-        response = _download_with_fallback_authn(config, url, data)
-        if response.ok:
-            destination_file.write(response.content)
-            logger.info(f'Completed {url}')
-            return response
+
+    if response is None or not response.ok:
+        if config.fallback_authn_enabled:
+            msg = ('No valid user access token in request or EDL OAuth authentication failed.'
+                   'Fallback authentication enabled: retrying with Basic auth.')
+            logger.warning(msg)
+            response = _download_with_fallback_authn(config, url, data)
+
+    if response.ok:
+        time_diff = datetime.datetime.now() - start_time
+        duration_ms = int(round(time_diff.total_seconds() * 1000))
+        destination_file.write(response.content)
```

Comment:
The file write seems lost in the middle of this timing work...could we extract the timing, regex extraction of host and url, and logging to a utility method/function?

Also, should the duration include the file write time? I wasn't sure if we intentionally get the duration before writing.

Do we want to log the timing in the fail cases too?

Reply:
Both of those were intentional (don't include the time it takes to write the file in logging the download time), and don't log the timings for failures because failures are likely to skew the timings and be unrelated to the download performance of the host.

I briefly looked at trying to put timing into a separate function, but things looked too intertwined at first. I can give it another try.

Reply:
I pushed a helper function that makes the main code a little bit cleaner in the successful download.